controllers/argocd/util.go

```diff
@@ -1360,6 +1368,88 @@ func (r *ReconcileArgoCD) setManagedNamespaces(cr *argoproj.ArgoCD) error {
 	return nil
 }
 
+func (r *ReconcileArgoCD) setManagedSourceNamespaces(cr *argoproj.ArgoCD) error {
+	r.ManagedSourceNamespaces = make(map[string]string)
+	namespaces := &corev1.NamespaceList{}
+	listOption := client.MatchingLabels{
+		common.ArgoCDManagedByClusterArgoCDLabel: cr.Namespace,
+	}
+
+	// get the list of namespaces managed by the Argo CD instance
+	if err := r.Client.List(context.TODO(), namespaces, listOption); err != nil {
+		return err
+	}
+
+	for _, namespace := range namespaces.Items {
+		r.ManagedSourceNamespaces[namespace.Name] = ""
+	}
+
+	return nil
+}
+
+// removeUnmanagedSourceNamespaceResources cleansup resources from SourceNamespaces if namespace is not managed by argocd instance.
+// It also removes the managed-by-cluster-argocd label from the namespace
+func (r *ReconcileArgoCD) removeUnmanagedSourceNamespaceResources(cr *argoproj.ArgoCD) error {
```

Comment:
I think we can simplify this function a little by providing the list of sourceNamespaces along with the cr as inputs 

Then the function just needs to delete the resources and remove the labels from those namespaces 
that might make this function a little more focused in scope and flexible in terms of where and how it is called

Reply:
Instead of taking the SourceNamespaces as an input, we are taking it directly from cr.spec.SourceNamespaces. So, the function will not be simplified as such.